src/OpenTK.Windowing.Desktop/NativeWindow.cs

```diff
@@ -953,29 +953,25 @@ private unsafe void ProcessInputEvents()
 
             for (var i = 0; i < _joystickStates.Length; i++)
             {
-                var joy = _joystickStates[i];
-                if (joy == default)
+                if (_joystickStates[i] == default)
                 {
                     continue;
                 }
 
-                var h = GLFW.GetJoystickHatsRaw(joy.Id, out var count);
-                var hats = new Hat[count];
+                var h = GLFW.GetJoystickHatsRaw(_joystickStates[i].Id, out var count);
```

Comment:
Here we want to change the GLFW bindings so we can pass it our own array and populate it ourselves, something like `GetJoystickHatsRaw(id, ref Hat[] values)`. `ref` so that the internal representation can call `Array.Resize` in necessary (the size will probably not change so resize will never actually be called, but good to have anyways). Then we would avoid all unnecessary allocations. The same refactoring should be done for the other Joystick input functions that return an array. It's is possible to do this refactoring because glfw gives us an unmanaged pointer at the lower level, so we can just copy from that into an existing array, rather than create on at that stage.

Reply:
This implies that one or other of `JoystickHats` and `Hats` needs to go away (which would be a sensible decision IMO)